db/write_controller.cc

```diff
@@ -62,49 +60,39 @@ uint64_t WriteController::GetDelay(SystemClock* clock, uint64_t num_bytes) {
   // interval.
   auto time_now = NowMicrosMonotonic(clock);
 
-  uint64_t sleep_debt = 0;
-  uint64_t time_since_last_refill = 0;
-  if (last_refill_time_ != 0) {
-    if (last_refill_time_ > time_now) {
-      sleep_debt = last_refill_time_ - time_now;
-    } else {
-      time_since_last_refill = time_now - last_refill_time_;
-      bytes_left_ +=
-          static_cast<uint64_t>(static_cast<double>(time_since_last_refill) /
-                                kMicrosPerSecond * delayed_write_rate_);
-      if (time_since_last_refill >= kRefillInterval &&
-          bytes_left_ > num_bytes) {
-        // If refill interval already passed and we have enough bytes
-        // return without extra sleeping.
-        last_refill_time_ = time_now;
-        bytes_left_ -= num_bytes;
-        return 0;
-      }
+  const uint64_t kMicrosPerSecond = 1000000;
+  const uint64_t kRefillInterval = 1024U;
+
+  if (next_refill_time_ == 0) {
+    // Start with an initial allotment of bytes for one interval
+    next_refill_time_ = time_now;
+  }
+  if (next_refill_time_ <= time_now) {
+    // Refill based on time interval plus any extra elapsed
+    uint64_t elapsed = time_now - next_refill_time_ + kRefillInterval;
+    bytes_left_ += static_cast<uint64_t>(
+        1.0 * elapsed / kMicrosPerSecond * delayed_write_rate_ + 0.999999);
+    next_refill_time_ = time_now + kRefillInterval;
```

Comment:
A couple of things here look wrong to me and could use an explanation.  time_now and next_refill_time are in time units but the refill interval is in bytes (1K).  This leads to a funky value for elapsed (is it bytes or time?).

Then a few lines later next_refill_time is also updated based on 1K, which is effectively upping the time by a millisecond I think.

Reply:
kRefillInterval is time everywhere. This hasn't changed. I don't know why it was chosen to be 1.024 ms, but I didn't find a bug in that so I didn't change it. I can add a clarifying comment.

The code author is unavailable atm.